[Quant][fx] Add lowering for Linear-Bn1d in QAT mode

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73509

Summary: This adds functionality to lower reference models
involving the Linear-Bn1d pattern in FX QAT mode. This follows
https://github.com/pytorch/pytorch/pull/72431 and https://github.com/pytorch/pytorch/pull/72796, which add Linear-Bn1d fusion functionality
to eager QAT mode.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_linear_module

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

Tasks: https://github.com/pytorch/pytorch/issues/66483

Differential Revision: [D34591251](https://our.internmc.facebook.com/intern/diff/D34591251)

andrewor14 committed:
[Quant][fx] Add lowering for Linear-Bn1d in QAT mode

Summary: This adds functionality to lower reference models
involving the Linear-Bn1d pattern in FX QAT mode.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_linear_module

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

Tasks: https://github.com/pytorch/pytorch/issues/66483

[ghstack-poisoned]

vkuzo commented on torch/nn/intrinsic/qat/modules/linear_fused.py:
```diff
@@ -1,11 +1,56 @@
 from torch.nn.parameter import Parameter
 
 
+class LinearReLU(nnqat.Linear, nni._FusedModule):
```
For larger PRs it might be good to separate unrelated changes (such as moving the `LinearReLU` code) into a separate PR to make things easier to review.  This PR is not that complicated so it's ok here, but just an optional suggestion for future PRs.

andrewor14 replied:
Ok, I'll keep that in mind

vkuzo reviewed (approved):
thanks for fixing this!

jerryzh168 reviewed (approved):
looks good!

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -3685,41 +3685,61 @@ def setUp(self):
-        for model, quant_type in options:
-            self.checkGraphModeFxOp(
-                model, data, quant_type, quantized_nodes[quant_type])
+        for quant_type in self.all_quant_types:
```
thanks, this refactor looks very clean

andrewor14 committed:
Update on "[Quant][fx] Add lowering for Linear-Bn1d in QAT mode"


Summary: This adds functionality to lower reference models
involving the Linear-Bn1d pattern in FX QAT mode. This follows
https://github.com/pytorch/pytorch/pull/72431 and https://github.com/pytorch/pytorch/pull/72796, which add Linear-Bn1d fusion functionality
to eager QAT mode.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_linear_module

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

Tasks: https://github.com/pytorch/pytorch/issues/66483

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Add lowering for Linear-Bn1d in QAT mode"


Summary: This adds functionality to lower reference models
involving the Linear-Bn1d pattern in FX QAT mode. This follows
https://github.com/pytorch/pytorch/pull/72431 and https://github.com/pytorch/pytorch/pull/72796, which add Linear-Bn1d fusion functionality
to eager QAT mode.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_linear_module

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

Tasks: https://github.com/pytorch/pytorch/issues/66483

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34591251).

andrewor14 committed:
Update on "[Quant][fx] Add lowering for Linear-Bn1d in QAT mode"


Summary: This adds functionality to lower reference models
involving the Linear-Bn1d pattern in FX QAT mode. This follows
https://github.com/pytorch/pytorch/pull/72431 and https://github.com/pytorch/pytorch/pull/72796, which add Linear-Bn1d fusion functionality
to eager QAT mode.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_linear_module

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

Tasks: https://github.com/pytorch/pytorch/issues/66483

Differential Revision: [D34591251](https://our.internmc.facebook.com/intern/diff/D34591251)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34591251).

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -221,7 +221,8 @@ def should_skip_lowering(op: torch.fx.node.Node, qconfig_map: Dict[str, QConfigA
 LOWER_FUSED_MODULE_MAP: Dict[Type[nn.Module], Tuple[Type[nn.Module], Type[WeightedQuantizedModule]]] = {
-    nni.LinearReLU: (nnqr.Linear, nniq.LinearReLU)
+    nni.LinearReLU: (nnqr.Linear, nniq.LinearReLU),
+    nni.LinearBn1d: (nnqr.Linear, nnq.Linear)
```
I feel we probably need to fuse bn into linear before the lowering step

jerryzh168 commented:
looks like we do not have a test comparing against eager mode here for linear bn fusion, that is why there is no test failing. maybe we could add a eager mode test, like https://github.com/pytorch/pytorch/blob/master/test/quantization/fx/test_quantize_fx.py#L955

andrewor14 committed:
Update on "[Quant][fx] Add lowering for Linear-Bn1d in QAT mode"


Summary: This adds functionality to lower reference models
involving the Linear-Bn1d pattern in FX QAT mode. This follows
https://github.com/pytorch/pytorch/pull/72431 and https://github.com/pytorch/pytorch/pull/72796, which add Linear-Bn1d fusion functionality
to eager QAT mode.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_linear_module

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

Tasks: https://github.com/pytorch/pytorch/issues/66483

Differential Revision: [D34591251](https://our.internmc.facebook.com/intern/diff/D34591251)

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Add lowering for Linear-Bn1d in QAT mode"


Summary: This adds functionality to lower reference models
involving the Linear-Bn1d pattern in FX QAT mode. This follows
https://github.com/pytorch/pytorch/pull/72431 and https://github.com/pytorch/pytorch/pull/72796, which add Linear-Bn1d fusion functionality
to eager QAT mode.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_linear_module

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

Tasks: https://github.com/pytorch/pytorch/issues/66483

Differential Revision: [D34591251](https://our.internmc.facebook.com/intern/diff/D34591251)

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Add lowering for Linear-Bn1d in QAT mode"


Summary: This adds functionality to lower reference models
involving the Linear-Bn1d pattern in FX QAT mode. This follows
https://github.com/pytorch/pytorch/pull/72431 and https://github.com/pytorch/pytorch/pull/72796, which add Linear-Bn1d fusion functionality
to eager QAT mode.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_linear_module

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

Tasks: https://github.com/pytorch/pytorch/issues/66483

Differential Revision: [D34591251](https://our.internmc.facebook.com/intern/diff/D34591251)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34591251).

andrewor14 committed:
Update on "[Quant][fx] Add lowering for Linear-Bn1d in QAT mode"


Summary: This adds functionality to lower reference models
involving the Linear-Bn1d pattern in FX QAT mode. This follows
https://github.com/pytorch/pytorch/pull/72431 and https://github.com/pytorch/pytorch/pull/72796, which add Linear-Bn1d fusion functionality
to eager QAT mode.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_linear_module

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

Tasks: https://github.com/pytorch/pytorch/issues/66483

Differential Revision: [D34591251](https://our.internmc.facebook.com/intern/diff/D34591251)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34591251).

andrewor14 committed:
Update on "[Quant][fx] Add lowering for Linear-Bn1d in QAT mode"


Summary: This adds functionality to lower reference models
involving the Linear-Bn1d pattern in FX QAT mode. This follows
https://github.com/pytorch/pytorch/pull/72431 and https://github.com/pytorch/pytorch/pull/72796, which add Linear-Bn1d fusion functionality
to eager QAT mode.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_linear_module

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

Tasks: https://github.com/pytorch/pytorch/issues/66483

Differential Revision: [D34591251](https://our.internmc.facebook.com/intern/diff/D34591251)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34591251).
